File: composeApp/src/main/kotlin/com/po4yka/trailglass/ui/permission/PermissionDialogs.kt
package com.po4yka.trailglass.ui.permission

import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.verticalScroll
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.CameraAlt
import androidx.compose.material.icons.filled.Check
import androidx.compose.material.icons.filled.Close
import androidx.compose.material.icons.filled.Info
import androidx.compose.material.icons.filled.LocationOn
import androidx.compose.material.icons.filled.Notifications
import androidx.compose.material.icons.filled.PhotoLibrary
import androidx.compose.material.icons.filled.Settings
import androidx.compose.material.icons.filled.Warning
import androidx.compose.material3.AlertDialog
import androidx.compose.material3.Button
import androidx.compose.material3.Icon
import androidx.compose.material3.IconButton
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Surface
import androidx.compose.material3.Text
import androidx.compose.material3.TextButton
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.vector.ImageVector
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import com.po4yka.trailglass.domain.permission.PermissionRequestState
import com.po4yka.trailglass.domain.permission.PermissionType
import com.po4yka.trailglass.domain.permission.SettingsInstructions

/**
 * Dialog showing permission rationale before requesting. Explains why the permission is needed and what features it
 * enables.
 */
@Composable
fun PermissionRationaleDialog(
    requestState: PermissionRequestState,
    onAccept: () -> Unit,
    onDeny: () -> Unit,
    modifier: Modifier = Modifier
) {
    AlertDialog(
        onDismissRequest = onDeny,
        icon = {
            Icon(
                imageVector = getPermissionIcon(requestState.permissionType),
                contentDescription = null,
                modifier = Modifier.size(32.dp),
                tint = MaterialTheme.colorScheme.primary
            )
        },
        title = {
            Text(
                text = requestState.rationale.title,
                style = MaterialTheme.typography.headlineSmall
            )
        },
        text = {
            Column(
                modifier = Modifier.verticalScroll(rememberScrollState()),
                verticalArrangement = Arrangement.spacedBy(16.dp)
            ) {
                // Description
                Text(
                    text = requestState.rationale.description,
                    style = MaterialTheme.typography.bodyMedium,
                    color = MaterialTheme.colorScheme.onSurfaceVariant
                )

                // Features list
                if (requestState.rationale.features.isNotEmpty()) {
                    Text(
                        text = "This permission enables:",
                        style = MaterialTheme.typography.titleSmall,
                        fontWeight = FontWeight.SemiBold
                    )

                    requestState.rationale.features.forEach { feature ->
                        Row(
                            modifier = Modifier.fillMaxWidth(),
                            horizontalArrangement = Arrangement.spacedBy(8.dp),
                            verticalAlignment = Alignment.Top
                        ) {
                            Icon(
                                imageVector = Icons.Default.Check,
                                contentDescription = null,
                                modifier = Modifier.size(20.dp),
                                tint = MaterialTheme.colorScheme.primary
                            )
                            Text(
                                text = feature,
                                style = MaterialTheme.typography.bodyMedium,
                                modifier = Modifier.weight(1f)
                            )
                        }
                    }
                }

                // Required/Optional indicator
                if (requestState.rationale.isRequired) {
                    Surface(
                        color = MaterialTheme.colorScheme.errorContainer,
                        shape = MaterialTheme.shapes.small
                    ) {
                        Row(
                            modifier = Modifier.padding(12.dp),
                            horizontalArrangement = Arrangement.spacedBy(8.dp),
                            verticalAlignment = Alignment.CenterVertically
                        ) {
                            Icon(
                                imageVector = Icons.Default.Info,
                                contentDescription = null,
                                modifier = Modifier.size(20.dp),
                                tint = MaterialTheme.colorScheme.onErrorContainer
                            )
                            Text(
                                text = "Required for core functionality",
                                style = MaterialTheme.typography.bodySmall,
                                color = MaterialTheme.colorScheme.onErrorContainer
                            )
                        }
                    }
                }
            }
        },
        confirmButton = {
            Button(onClick = onAccept) {
                Text("Grant Permission")
            }
        },
        dismissButton = {
            TextButton(onClick = onDeny) {
                Text(if (requestState.rationale.isRequired) "Not Now" else "Skip")
            }
        },
        modifier = modifier
    )
}

/** Dialog shown when permission is denied (but not permanently). Encourages user to reconsider. */
@Composable
fun PermissionDeniedDialog(
    requestState: PermissionRequestState,
    onRetry: () -> Unit,
    onContinue: () -> Unit,
    modifier: Modifier = Modifier
) {
    AlertDialog(
        onDismissRequest = onContinue,
        icon = {
            Icon(
                imageVector = Icons.Default.Warning,
                contentDescription = null,
                modifier = Modifier.size(32.dp),
                tint = MaterialTheme.colorScheme.error
            )
        },
        title = {
            Text("Permission Denied")
        },
        text = {
            Column(verticalArrangement = Arrangement.spacedBy(12.dp)) {
                Text(
                    text = "Without ${requestState.rationale.title.lowercase()}, the following features won't work:",
                    style = MaterialTheme.typography.bodyMedium
                )

                requestState.rationale.features.take(3).forEach { feature ->
                    Row(
                        horizontalArrangement = Arrangement.spacedBy(8.dp),
                        verticalAlignment = Alignment.Top
                    ) {
                        Icon(
                            imageVector = Icons.Default.Close,
                            contentDescription = null,
                            modifier = Modifier.size(20.dp),
                            tint = MaterialTheme.colorScheme.error
                        )
                        Text(
                            text = feature,
                            style = MaterialTheme.typography.bodySmall,
                            modifier = Modifier.weight(1f)
                        )
                    }
                }

                if (requestState.rationale.isRequired) {
                    Text(
                        text = "This permission is required for TrailGlass to function properly.",
                        style = MaterialTheme.typography.bodySmall,
                        color = MaterialTheme.colorScheme.error,
                        fontWeight = FontWeight.SemiBold
                    )
                }
            }
        },
        confirmButton = {
            Button(onClick = onRetry) {
                Text("Try Again")
            }
        },
        dismissButton = {
            if (!requestState.rationale.isRequired) {
                TextButton(onClick = onContinue) {
                    Text("Continue Anyway")
                }
            }
        },
        modifier = modifier
    )
}

/** Dialog shown when permission is permanently denied. Provides instructions to enable in settings. */
@Composable
fun PermissionPermanentlyDeniedDialog(
    requestState: PermissionRequestState,
    instructions: SettingsInstructions,
    onOpenSettings: () -> Unit,
    onDismiss: () -> Unit,
    modifier: Modifier = Modifier
) {
    AlertDialog(
        onDismissRequest = onDismiss,
        icon = {
            Icon(
                imageVector = Icons.Default.Settings,
                contentDescription = null,
                modifier = Modifier.size(32.dp),
                tint = MaterialTheme.colorScheme.primary
            )
        },
        title = {
            Text("Permission Required")
        },
        text = {
            Column(
                modifier = Modifier.verticalScroll(rememberScrollState()),
                verticalArrangement = Arrangement.spacedBy(16.dp)
            ) {
                Text(
                    text =
                        "You've previously denied ${requestState.rationale.title.lowercase()}. " +
                            "To use this feature, you'll need to enable it in Settings.",
                    style = MaterialTheme.typography.bodyMedium
                )

                // Instructions
                Surface(
                    color = MaterialTheme.colorScheme.surfaceVariant,
                    shape = MaterialTheme.shapes.medium
                ) {
                    Column(
                        modifier = Modifier.padding(16.dp),
                        verticalArrangement = Arrangement.spacedBy(8.dp)
                    ) {
                        Text(
                            text = "How to enable:",
                            style = MaterialTheme.typography.titleSmall,
                            fontWeight = FontWeight.SemiBold
                        )

                        instructions.steps.forEachIndexed { index, step ->
                            Row(
                                horizontalArrangement = Arrangement.spacedBy(12.dp),
                                verticalAlignment = Alignment.Top
                            ) {
                                Surface(
                                    color = MaterialTheme.colorScheme.primary,
                                    shape = MaterialTheme.shapes.small,
                                    modifier = Modifier.size(24.dp)
                                ) {
                                    Box(contentAlignment = Alignment.Center) {
                                        Text(
                                            text = "${index + 1}",
                                            style = MaterialTheme.typography.labelSmall,
                                            color = MaterialTheme.colorScheme.onPrimary
                                        )
                                    }
                                }
                                Text(
                                    text = step,
                                    style = MaterialTheme.typography.bodyMedium,
                                    modifier = Modifier.weight(1f)
                                )
                            }
                        }
                    }
                }
            }
        },
        confirmButton = {
            Button(onClick = onOpenSettings) {
                Icon(
                    imageVector = Icons.Default.Settings,
                    contentDescription = null,
                    modifier = Modifier.size(18.dp)
                )
                Spacer(Modifier.width(8.dp))
                Text("Open Settings")
            }
        },
        dismissButton = {
            if (!requestState.rationale.isRequired) {
                TextButton(onClick = onDismiss) {
                    Text("Maybe Later")
                }
            }
        },
        modifier = modifier
    )
}

/** Compact permission request banner (alternative to full dialog). Shows at top of screen with brief explanation. */
@Composable
fun PermissionRequestBanner(
    requestState: PermissionRequestState,
    onGrant: () -> Unit,
    onDismiss: () -> Unit,
    modifier: Modifier = Modifier
) {
    Surface(
        color = MaterialTheme.colorScheme.primaryContainer,
        modifier = modifier.fillMaxWidth()
    ) {
        Row(
            modifier = Modifier.padding(16.dp),
            horizontalArrangement = Arrangement.spacedBy(12.dp),
            verticalAlignment = Alignment.CenterVertically
        ) {
            Icon(
                imageVector = getPermissionIcon(requestState.permissionType),
                contentDescription = null,
                tint = MaterialTheme.colorScheme.onPrimaryContainer
            )

            Column(
                modifier = Modifier.weight(1f),
                verticalArrangement = Arrangement.spacedBy(4.dp)
            ) {
                Text(
                    text = requestState.rationale.title,
                    style = MaterialTheme.typography.titleSmall,
                    fontWeight = FontWeight.SemiBold,
                    color = MaterialTheme.colorScheme.onPrimaryContainer
                )
                Text(
                    text = requestState.rationale.description,
                    style = MaterialTheme.typography.bodySmall,
                    color = MaterialTheme.colorScheme.onPrimaryContainer
                )
            }

            TextButton(onClick = onGrant) {
                Text("Grant")
            }

            IconButton(onClick = onDismiss) {
                Icon(
                    imageVector = Icons.Default.Close,
                    contentDescription = "Dismiss",
                    tint = MaterialTheme.colorScheme.onPrimaryContainer
                )
            }
        }
    }
}

/** Get icon for permission type. */
private fun getPermissionIcon(permissionType: PermissionType): ImageVector =
    when (permissionType) {
        PermissionType.LOCATION_FINE,
        PermissionType.LOCATION_COARSE,
        PermissionType.LOCATION_BACKGROUND -> Icons.Default.LocationOn

        PermissionType.CAMERA -> Icons.Default.CameraAlt
        PermissionType.PHOTO_LIBRARY -> Icons.Default.PhotoLibrary
        PermissionType.NOTIFICATIONS -> Icons.Default.Notifications
    }

File: iosApp/iosApp/Permission/PermissionDialogs.swift
import SwiftUI
import Shared

/// SwiftUI view showing permission rationale before requesting.
struct PermissionRationaleView: View {
    let requestState: PermissionRequestState
    let onAccept: () -> Void
    let onDeny: () -> Void

    @Environment(\.dismiss) private var dismiss

    var body: some View {
        NavigationView {
            ScrollView {
                VStack(alignment: .leading, spacing: 24) {
                    // Icon and title
                    HStack(spacing: 16) {
                        Image(systemName: permissionIcon)
                            .font(.system(size: 40))
                            .foregroundColor(.adaptivePrimary)

                        Text(requestState.rationale.title)
                            .font(.title2)
                            .fontWeight(.bold)
                    }

                    // Description
                    Text(requestState.rationale.description)
                        .font(.body)
                        .foregroundColor(.secondary)

                    // Features list
                    if !requestState.rationale.features.isEmpty {
                        VStack(alignment: .leading, spacing: 12) {
                            Text("This permission enables:")
                                .font(.headline)

                            ForEach(requestState.rationale.features, id: \.self) { feature in
                                HStack(alignment: .top, spacing: 12) {
                                    Image(systemName: "checkmark.circle.fill")
                                        .foregroundColor(.adaptiveSuccess)
                                        .font(.system(size: 20))

                                    Text(feature)
                                        .font(.body)
                                        .fixedSize(horizontal: false, vertical: true)
                                }
                            }
                        }
                    }

                    // Required indicator
                    if requestState.rationale.isRequired {
                        HStack(spacing: 12) {
                            Image(systemName: "exclamationmark.triangle.fill")
                                .foregroundColor(.adaptiveWarning)

                            Text("Required for core functionality")
                                .font(.caption)
                                .fontWeight(.semibold)
                        }
                        .padding()
                        .frame(maxWidth: .infinity, alignment: .leading)
                        .background(Color.adaptiveWarning.opacity(0.1))
                        .cornerRadius(8)
                    }

                    Spacer()

                    // Buttons
                    VStack(spacing: 12) {
                        Button(action: onAccept) {
                            Text("Grant Permission")
                                .fontWeight(.semibold)
                                .frame(maxWidth: .infinity)
                                .padding()
                                .background(Color.adaptivePrimary)
                                .foregroundColor(.white)
                                .cornerRadius(12)
                        }

                        Button(action: {
                            onDeny()
                            dismiss()
                        }) {
                            Text(requestState.rationale.isRequired ? "Not Now" : "Skip")
                                .foregroundColor(.secondary)
                        }
                    }
                }
                .padding()
            }
            .navigationBarTitleDisplayMode(.inline)
            .toolbar {
                ToolbarItem(placement: .navigationBarTrailing) {
                    Button(action: {
                        onDeny()
                        dismiss()
                    }) {
                        Image(systemName: "xmark.circle.fill")
                            .foregroundColor(.secondary)
                    }
                }
            }
        }
    }

    private var permissionIcon: String {
        switch requestState.permissionType {
        case .locationFine, .locationCoarse, .locationBackground:
            return "location.circle.fill"
        case .camera:
            return "camera.fill"
        case .photoLibrary:
            return "photo.fill"
        case .notifications:
            return "bell.fill"
        default:
            return "questionmark.circle.fill"
        }
    }
}

/// Alert view showing permission denied message.
struct PermissionDeniedAlert: View {
    let requestState: PermissionRequestState
    let onRetry: () -> Void
    let onContinue: () -> Void

    var body: some View {
        VStack(spacing: 20) {
            Image(systemName: "exclamationmark.triangle.fill")
                .font(.system(size: 50))
                .foregroundColor(.adaptiveWarning)

            Text("Permission Denied")
                .font(.title2)
                .fontWeight(.bold)

            Text("Without \(requestState.rationale.title.lowercased()), the following features won't work:")
                .font(.body)
                .multilineTextAlignment(.center)
                .foregroundColor(.secondary)

            VStack(alignment: .leading, spacing: 8) {
                ForEach(Array(requestState.rationale.features.prefix(3)), id: \.self) { feature in
                    HStack(alignment: .top, spacing: 8) {
                        Image(systemName: "xmark.circle.fill")
                            .foregroundColor(.adaptiveWarning)
                        Text(feature)
                            .font(.caption)
                            .fixedSize(horizontal: false, vertical: true)
                    }
                }
            }
            .padding()
            .background(Color(.systemGray6))
            .cornerRadius(8)

            if requestState.rationale.isRequired {
                Text("This permission is required for TrailGlass to function properly.")
                    .font(.caption)
                    .fontWeight(.semibold)
                    .foregroundColor(.adaptiveWarning)
                    .multilineTextAlignment(.center)
            }

            VStack(spacing: 12) {
                Button(action: onRetry) {
                    Text("Try Again")
                        .fontWeight(.semibold)
                        .frame(maxWidth: .infinity)
                        .padding()
                        .background(Color.adaptivePrimary)
                        .foregroundColor(.white)
                        .cornerRadius(12)
                }

                if !requestState.rationale.isRequired {
                    Button(action: onContinue) {
                        Text("Continue Anyway")
                            .foregroundColor(.secondary)
                    }
                }
            }
        }
        .padding()
    }
}

/// View showing instructions to enable permission in Settings.
struct PermissionSettingsInstructionsView: View {
    let requestState: PermissionRequestState
    let instructions: SettingsInstructions
    let onOpenSettings: () -> Void
    let onDismiss: () -> Void

    @Environment(\.dismiss) private var dismiss

    var body: some View {
        NavigationView {
            ScrollView {
                VStack(alignment: .leading, spacing: 24) {
                    // Header
                    VStack(alignment: .leading, spacing: 8) {
                        Image(systemName: "gear.circle.fill")
                            .font(.system(size: 50))
                            .foregroundColor(.adaptivePrimary)

                        Text("Permission Required")
                            .font(.title)
                            .fontWeight(.bold)

                        Text("You've previously denied \(requestState.rationale.title.lowercased()). To use this feature, you'll need to enable it in Settings.")
                            .font(.body)
                            .foregroundColor(.secondary)
                    }

                    // Instructions
                    VStack(alignment: .leading, spacing: 16) {
                        Text("How to enable:")
                            .font(.headline)

                        VStack(alignment: .leading, spacing: 12) {
                            ForEach(Array(instructions.steps.enumerated()), id: \.offset) { index, step in
                                HStack(alignment: .top, spacing: 16) {
                                    ZStack {
                                        Circle()
                                            .fill(Color.adaptivePrimary)
                                            .frame(width: 28, height: 28)

                                        Text("\(index + 1)")
                                            .font(.caption)
                                            .fontWeight(.bold)
                                            .foregroundColor(.white)
                                    }

                                    Text(step)
                                        .font(.body)
                                        .fixedSize(horizontal: false, vertical: true)
                                }
                            }
                        }
                        .padding()
                        .background(Color(.systemGray6))
                        .cornerRadius(12)
                    }

                    Spacer()

                    // Buttons
                    VStack(spacing: 12) {
                        Button(action: onOpenSettings) {
                            HStack {
                                Image(systemName: "gear")
                                Text("Open Settings")
                            }
                            .fontWeight(.semibold)
                            .frame(maxWidth: .infinity)
                            .padding()
                            .background(Color.adaptivePrimary)
                            .foregroundColor(.white)
                            .cornerRadius(12)
                        }

                        if !requestState.rationale.isRequired {
                            Button(action: {
                                onDismiss()
                                dismiss()
                            }) {
                                Text("Maybe Later")
                                    .foregroundColor(.secondary)
                            }
                        }
                    }
                }
                .padding()
            }
            .navigationBarTitleDisplayMode(.inline)
            .toolbar {
                ToolbarItem(placement: .navigationBarTrailing) {
                    if !requestState.rationale.isRequired {
                        Button(action: {
                            onDismiss()
                            dismiss()
                        }) {
                            Image(systemName: "xmark.circle.fill")
                                .foregroundColor(.secondary)
                        }
                    }
                }
            }
        }
    }
}

/// Compact permission banner for inline display.
struct PermissionRequestBanner: View {
    let requestState: PermissionRequestState
    let onGrant: () -> Void
    let onDismiss: () -> Void

    var body: some View {
        HStack(spacing: 12) {
            Image(systemName: permissionIcon)
                .font(.system(size: 24))
                .foregroundColor(.adaptivePrimary)

            VStack(alignment: .leading, spacing: 4) {
                Text(requestState.rationale.title)
                    .font(.subheadline)
                    .fontWeight(.semibold)

                Text(requestState.rationale.description)
                    .font(.caption)
                    .foregroundColor(.secondary)
                    .lineLimit(2)
            }

            Spacer()

            Button(action: onGrant) {
                Text("Grant")
                    .font(.caption)
                    .fontWeight(.semibold)
                    .padding(.horizontal, 12)
                    .padding(.vertical, 6)
                    .background(Color.adaptivePrimary)
                    .foregroundColor(.white)
                    .cornerRadius(8)
            }

            Button(action: onDismiss) {
                Image(systemName: "xmark.circle.fill")
                    .foregroundColor(.secondary)
            }
        }
        .padding()
        .background(Color.adaptivePrimary.opacity(0.1))
        .cornerRadius(12)
        .padding(.horizontal)
    }

    private var permissionIcon: String {
        switch requestState.permissionType {
        case .locationFine, .locationCoarse, .locationBackground:
            return "location.circle.fill"
        case .camera:
            return "camera.fill"
        case .photoLibrary:
            return "photo.fill"
        case .notifications:
            return "bell.fill"
        default:
            return "questionmark.circle.fill"
        }
    }
}

// MARK: - Preview
#if DEBUG
struct PermissionDialogs_Previews: PreviewProvider {
    static var previews: some View {
        let mockRationale = PermissionRationale(
            permissionType: .locationFine,
            title: "Precise Location Access",
            description: "TrailGlass needs precise location access to automatically record your visits.",
            features: [
                "Automatically detect when you arrive at places",
                "Accurately map your routes",
                "Calculate distances and durations"
            ],
            isRequired: true
        )

        let mockState = PermissionRequestState(
            permissionType: .locationFine,
            state: .notDetermined,
            rationale: mockRationale,
            shouldShowRationale: true,
            canRequest: true
        )

        let mockInstructions = SettingsInstructions(
            permissionType: .locationFine,
            steps: [
                "Open Settings app",
                "Find and tap TrailGlass",
                "Tap Location",
                "Select 'While Using the App' or 'Always'"
            ],
            quickAction: "Open Settings"
        )

        Group {
            PermissionRationaleView(
                requestState: mockState,
                onAccept: {},
                onDeny: {}
            )
            .previewDisplayName("Rationale")

            PermissionSettingsInstructionsView(
                requestState: mockState,
                instructions: mockInstructions,
                onOpenSettings: {},
                onDismiss: {}
            )
            .previewDisplayName("Settings Instructions")

            PermissionRequestBanner(
                requestState: mockState,
                onGrant: {},
                onDismiss: {}
            )
            .previewDisplayName("Banner")
        }
    }
}
#endif
